### PR TITLE
Fix charset parsing in HttpListenerRequest

### PR DESCRIPTION
**Pull request by caesar1995**
The function `GetAttributeFromHeader` was ported from Framework. It's obsolete and serves as a single purpose to parse the `charset` attribute for `ContentType` header in HttpListener's request.

If the parameter value ends with a `;`, after parsing, the value will contain that undesired semicolon in it. 
For example, `"charset=unicode;"` will be parsed to `"unicode;"`, thus the following `Encoding.GetEncoding(charSet)` will fail with `ArgumentException`. In this case, `Encoding.Default` will be chose for `ContentEncoding`, even if client explicitly specifies the Unicode encoding in the header.

The fix is following rfc2045 section 5.1.:

```
5.1.  Syntax of the Content-Type Header Field

     content := "Content-Type" ":" type "/" subtype
                *(";" parameter)

parameter := attribute "=" value
```

Usually `charset` parameter is at the end, and we don't need to worry about the `;`. But it could be possible that `charset` parameter is in the middle, and followed by a `;`, thus fails parsing.

Per RFC, the last parameter should not followed by a `;`. In our code, the parsing is more robust, which I think it's acceptable. Please let me know if you have different thoughts.

Close: #23506

### CODE REVIEW

File: src/System.Net.HttpListener/tests/HttpListenerRequestTests.cs [hunk 1]
```diff
@@ -82,7 +82,24 @@ public static IEnumerable<object[]> ContentEncoding_TestData()
             yield return new object[] { "Content-Type:application.json;charset=,", Encoding.Default };
             yield return new object[] { "Content-Type:application.json;charset=\"unicode", Encoding.Default };
             yield return new object[] { "Content-Type:application/json;charset=NoSuchEncoding", Encoding.Default };
-            yield return new object[] { "Content-Type:application/json;charset=unicode; boundary=something", Encoding.Default };
```
Comment (caesar1995) [inline src/System.Net.HttpListener/tests/HttpListenerRequestTests.cs]: The test data is not correct and I changed it.

Reason:
Without the change, the `charset=unicode;` will be parsed as `unicode;` (notice the semicolon inside the value), thus the Encoding.GetEncoding(charSet) will fail with ArgumentException. In this case, Encoding.Default will be chose for ContentEncoding, even if client explicitly specifies the Unicode encoding in the header.

**Comment by caesar1995** [comment]
The tests I added demonstrate the same bug in .NET Framework. i.e., `charset=unicode;` will be parsed as `unicode;`, for the same reason, `Encoding.Default` is chose.

We will need to fix Framework issue as well. For now, I will disable the failing test data.

File: src/System.Net.HttpListener/src/System/Net/HttpListenerRequest.cs [hunk 2]
```diff
@@ -290,8 +290,10 @@ private static class Helpers
             //
             // Get attribute off header value
             //
-            internal static string GetAttributeFromHeader(string headerValue, string attrName)
+            internal static string GetCharSetValueFromHeader(string headerValue)
             {
+                string attrName = "charset";
```
Comment (stephentoub) [inline src/System.Net.HttpListener/src/System/Net/HttpListenerRequest.cs]: Nit:
```C#
const string AttrName = "charset";
```

Comment (caesar1995) [reply src/System.Net.HttpListener/src/System/Net/HttpListenerRequest.cs]: Fixed.

**Pull request merged**